modules/core/src/main/scala/com/azavea/stac4s/extensions/CatalogExtension.scala

```diff
@@ -0,0 +1,25 @@
+package com.azavea.stac4s.extensions
+
+import com.azavea.stac4s.StacCatalog
+
+import io.circe.{Decoder, Encoder}
+import io.circe.syntax._
+
+trait CatalogExtension[T] {
+  def getProperties(catalog: StacCatalog): ExtensionResult[T]
+  def extend(catalog: StacCatalog, extensionFields: T): StacCatalog
```

Comment:
The issue with `addProperties` is that for some things that means "add to the `properties` field" (items) while for everything else it means "add top level keys". I don't know whether that matters to consumers but I spent a lot of time looking at extensions while working on this and wouldn't want to accidentally pun on how an extension modifies an object. It's tough for me to tell whether that's important or an artifact of being in the different extension spec for an hour.

I didn't think the collision with `extends` as in traits / classes was a big deal because:

- it's a function, so it can't ever appear in that position
- it refers to `extension` pretty transparently

Reply:
Oh, I see; makes sense. Should / can Items have extensionFields as well?

Reply:
Should not also be `getProperties` be smth like `getExtensionFIelds`? 
I really though about `.properties` field here by a mistake.

Reply:
When items have extensions, the extensions (at least in practice) seem always to go on `properties`, which is already a free-form `JsonObject`. It would make sense if `properties` on items had some consistent required shape, but since they're allowed to be [literally anything](https://github.com/radiantearth/stac-spec/blob/master/item-spec/item-spec.md#properties-object) and since extensions don't tend to mess with the top-level keys of items it's different for them.

Reply:
@jamesmcclain Btw collections can have a `property` extension as well; would not getPropertoes be confusing in this case for `Collections`?

Reply:
@jisantuc Btw collections can have a property extension as well; would not getPropertoes be confusing in this case for Collections?

Reply:
Yeah it's a mess... not only do they have `properties`, but their extensions add top-level fields. I'll rename, you make a lot of good points.

Reply:
75a67f9 and 680f399